### DISCUSSION TITLE
Incorrect Order of Arithmetic Operations

### DISCUSSION THREAD

**Discussion by piotrbajdek** [red/red#5277] [Q&A] [answered]
**Describe the bug**
Incorrect Order of Arithmetic Operations

**To reproduce**

>> 2 + 2 * 10
== 40
>> 2 + 2 ** 10
== 1048576
>> 2 * 2 ** 10
== 1048576

**Expected behavior**
Should be:
>> 2 + 2 * 10
== 22
>> 2 + 2 ** 10
== 1026
>> 2 * 2 ** 10
== 2048

**Platform version**
RED: [ branch: "master" tag: #v0.6.4 ahead: 4629 date: 28-Jan-2023/11:28:13 commit: #ec0c31279023590e285fb2e1869380670ee0239f ]
PLATFORM: [ name: "Fedora Linux 37 (Container Image)" OS: 'Linux arch: 'x86_64 version: 6.1.7 build: {#1 SMP PREEMPT_DYNAMIC Wed Jan 18 17:11:49 UTC 2023} ]

**Comment by hiiamboris** [comment, marked as answer]
Quoting the [R2 core manual](http://www.rebol.com/docs/core23/rebolcore-11.html#section-3) which is also relevant for Red here:
> The evaluation of expressions from left to right is independent of the type of operator that is used. For example:
> ```
> print 1 + 2 * 3
> 9
> ```

And:
> In REBOL, rather than requiring users to remember the precedence of operators, you only need to remember the left-to-right rule. More importantly, for advanced code such as expressions that handle expressions (in reflection, for example) you do not need to reorder terms based on precedence. The evaluation order is kept simple.

My personal view is this design just makes the interpreter much faster and simpler, as this way it operates on the data directly, not requiring an [AST](https://en.wikipedia.org/wiki/Abstract_syntax_tree).

When you want the commonly used order you can use `math` function:
```
>> math [2 + 2 * 10]
== 22
>> math [2 + 2 ** 10]
== 1026
>> math [2 * 2 ** 10]
== 2048
```

**Reply by piotrbajdek** [reply]
Thank you for your prompt response, @hiiamboris ! I comprehend that this design decision streamlines the interpreter. Nevertheless, it is mathematically flawed, regardless of whether it was inherited from Rebol or not. Thank you for pointing out the workaround using the `math` function. As a newcomer to Red, I will ensure to employ it in my future work. 😊

**Reply by greggirwin** [reply]
Remember that Red is a programming language, so does not *have to* adhere to standard math rules when evaluating expressions that include infix operators. Also, be aware that you can define new infix operators, which would be massively complicated both for definitions and for sharing user-defined operators. That's very important in the realm of dialects.

**Reply by piotrbajdek** [reply]
> Remember that Red is a programming language, so does not _have to_ adhere to standard math rules when evaluating expressions that include infix operators.

I understand the reasoning behind this design decision, yet it is the first programming language that deviates from standard mathematical conventions that I have encountered. This is quite _peculiar_.

> Also, be aware that you can define new infix operators, which would be massively complicated both for definitions and for sharing user-defined operators. That's very important in the realm of dialects.

This sounds quite impressive.

**Reply by hiiamboris** [reply]
When I first saw REBOL discarding math rules, I thought "what a weird decision". But quite quickly I got used to it and even found it to often be more convenient in practice. Instead of adding parentheses I could just rewrite the expression order. And when I inspected my code, I also found that visually following the expression left to right is simpler than building an AST in my mind and then trying to interpret it and predict the outcome.

So my point: don't worry about it. In the long run it's more a blessing than a curse.

**Reply by piotrbajdek** [reply]
I apologize if it seemed like I was attacking the design decisions. My intention was simply to report what I thought was a bug. I am currently focused on my Rust projects, but I am still interested in experimenting with Red as it appears promising. However, for mathematical purposes, I believe that languages that strictly adhere to mathematical conventions are less prone to bugs, so I would likely choose Julia or Fortran.

**Reply by greggirwin** [reply]
No offense taken. Many things in Red seems strange to newcomers. There are always design tradeoffs but, having used Redbol langs for 20+ years, I find other languages terribly limiting and archaic most of the time. It's always fair to ask "Why does it work that way?" and some of the answers to what people think are bugs often bend their minds. :^) Ultimately, Red is a great language construction kit, as `math` shows how to build a tiny DSL for that domain.

**Reply by piotrbajdek** [reply]
> There are always design tradeoffs but, having used Redbol langs for 20+ years, I find other languages terribly limiting and archaic most of the time.

But you're still on 32 bits! Just joking! 😜

**Reply by hiiamboris** [reply]
Yes, 32 bitness, and for serious math a lot is left to be desired. Also alpha stage of development. These are fair points of critique at the moment.

**Reply by piotrbajdek** [reply]
Is the Discussions section the appropriate place to ask questions about Red, or would you prefer that I inquire elsewhere?

**Reply by hiiamboris** [reply]
Ideally, [there](https://gitter.im/red/help) ☻

**Comment by dockimbel** [comment]
@piotrbajdek 

A few comments:
* The cognitive overload of dozen(s) of precedence rules with sometimes exception(s) is _immense_ compared to the simple left-to-right rule. It takes little time to get used to it as it is becomes quickly intuitive. I think that no Red/Rebol coder would want to come back the [operators precedence-hell](https://en.cppreference.com/w/c/language/operator_precedence). Fortran also has its [own unintuitive exceptions](https://docs.oracle.com/cd/E19957-01/805-4939/6j4m0vn72/index.html) for operators precedence.
* I would argue that it is easier to make mistakes when using precedence-base rules than the single-rule approach we use.
* There are mainstream languages like Forth or Postscript that do not use precedence rules for operators but rather [postfix](https://en.wikipedia.org/wiki/Reverse_Polish_notation) [notation](https://wiki.c2.com/?PostfixNotation). Math teachers, scientists and financial analysts since the 70's often rely on _postfix-notation_ calculators like the HP 11C or [12C](https://www.youtube.com/watch?v=bdTkKKz3MFc), which are both faster and easier to use than classic ones requiring parentheses to mark precedence.
